includes/distributor-customizations/class-author-ingestion.php

```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * Newspack Network author ingestino.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack_Network\Distributor_Customizations;
+
+use Newspack\Data_Events;
+use Newspack_Network\Debugger;
+use Newspack_Network\User_Update_Watcher;
+use Newspack_Network\Utils\Users as User_Utils;
+
+/**
+ * Class to handle author ingestion.
+ *
+ * This class is used to handle the authorship data added by Author_Distriburion to the distributed post
+ */
+class Author_Ingestion {
+
+	/**
+	 * Array of authors from pulled posts. See self::capture_authorship.
+	 *
+	 * @var array
+	 */
+	private static $pulled_posts = [];
+
+	/**
+	 * Initializes the class
+	 *
+	 * @return void
+	 */
+	public static function init() {
+		add_action( 'rest_insert_post', [ __CLASS__, 'handle_rest_insertion' ], 10, 2 );
+		add_filter( 'dt_item_mapping', [ __CLASS__, 'capture_authorship' ], 10, 2 );
+		add_action( 'dt_pull_post', [ __CLASS__, 'handle_pull' ] );
+	}
+
+	/**
+	 * Gets the CoAuthors Plus main object, if present
+	 *
+	 * @return false|\CoAuthors_Plus
+	 */
+	public static function get_coauthors_plus() {
+		global $coauthors_plus;
+		if ( is_null( $coauthors_plus ) || ( ! $coauthors_plus instanceof \CoAuthors_Plus ) ) {
+			return false;
+		}
+		return $coauthors_plus;
+	}
+
+	/**
+	 * Fired when a post is inserted via the REST API
+	 *
+	 * Checks if there are newspack network author information and ingests it.
+	 *
+	 * @param WP_Post         $post     Inserted post object.
+	 * @param WP_REST_Request $request  Request object.
+	 * @return void
+	 */
+	public static function handle_rest_insertion( $post, $request ) {
+
+		$distributed_authors = $request->get_param( 'newspack_network_authors' );
+		if ( empty( $distributed_authors ) ) {
+			return;
+		}
+
+		self::ingest_authors_for_post( $post->ID, $distributed_authors );
+	}
+
+	/**
+	 * Ingest authors for a post distribubted to this site
+	 *
+	 * @param int   $post_id The post ID.
+	 * @param array $distributed_authors The distributed authors array.
+	 * @return void
+	 */
+	public static function ingest_authors_for_post( $post_id, $distributed_authors ) {
+
+		Debugger::log( 'Ingesting authors from distributed post.' );
+
+		update_post_meta( $post_id, 'newspack_network_authors', $distributed_authors );
+
+		$coauthors_plus = self::get_coauthors_plus();
+		$coauthors      = [];
+
+		foreach ( $distributed_authors as $author ) {
+			// We only ingest WP Users. Guest authors are only stored in the newspack_network_authors post meta.
+			if ( empty( $author['type'] ) || 'wp_user' != $author['type'] ) {
+				continue;
+			}
+
+			Debugger::log( 'Ingesting author: ' . $author['user_email'] );
+
+			$user = User_Utils::get_or_create_user_by_email( $author['user_email'], get_post_meta( $post_id, 'dt_original_site_url', true ), $author['id'] );
```

Comment:
The author id. Why confusing? When we create a user that has been propagated by another site we store their original ID

Reply:
Because according to [the code inline docs](https://github.com/Automattic/newspack-network/blob/7e8e8271b2ab3dc6237199b899f254c4b12885b5/includes/utils/class-users.php#L20):

> @param string $remote_site_id The ID of the remote site. Used only when a new user is created.

Reply:
ah, I'll fix the docs. It should be The id ON the remote site... I'll make it clearer

Reply:
Done in https://github.com/Automattic/newspack-network/pull/17/commits/92c6b321026d0d1af4d47af04ca939927341b86e